Feat-2 / 친구 요청 기능

anewsdev opened the pull request:
- 제목 : Feat: Add Friend Request

## 🔘Part

- 친구 요청 파트

## 🔎 작업 내용

- Friend 관련 entity, dto, enum, controller, service, repository 추가
- 친구 요청 기능 추가 (Postman 테스트 완료)


## 🔧 앞으로의 과제

- 친구 수락 > 조회 > 삭제 > 거절 순으로 진행 예정

## ➕ 이슈 링크
- 없음

anewsdev committed:
init : Add entity, enum, dto, controller, service, repository create
 - entity : 친구, 친구 요청 entity 생성
 - enum : 친구 상태 확인을 위한 enum 추가
 - dto :  요청 및 반환 DTO Class responseDto, requestDto 추가
 - controller : FriendContoller 추가
 - service : FriendService 추가
 - repository : FriendRepository 추가

anewsdev committed:
feat : Add Friend Request
- 로그인한 유저의 친구 요청 기능 추가

seongjun1130 commented on src/main/java/com/sparta/newsfeedproject/domain/friend/service/FriendService.java:
```diff
@@ -0,0 +1,32 @@
+    private final FriendRequestRepository friendRequestRepository;
+    private final MemberRepository memberRepository;
+
+    public void sendFriendRequest(Long requesterId, Long receiverId) {
+        Member requester = memberRepository.findById(requesterId)
+                .orElseThrow(() -> new IllegalArgumentException("요청 유저를 찾을 수 없습니다."));
+        Member receiver = memberRepository.findById(receiverId)
+                .orElseThrow(()-> new IllegalArgumentException("수신 유저를 찾을 수 없습니다."));
```
1. IllegalArgumentException 말고 CustomException 을 통한 예외로 전환하시면 좋을거같습니다.
2. 동일한 관계의 신청 Ex: 1->2 And 2->1 일때 등 파생적 예외가 발생할 경우가 보입니다. (이미 친구상태일때는?)

anewsdev replied:
피드백 적용 완료 입니다.

seongjun1130 commented on src/main/java/com/sparta/newsfeedproject/domain/friend/service/FriendService.java:
```diff
@@ -0,0 +1,32 @@
+        Member receiver = memberRepository.findById(receiverId)
+                .orElseThrow(()-> new IllegalArgumentException("수신 유저를 찾을 수 없습니다."));
+
+        FriendRequest friendRequest = new FriendRequest();
+        friendRequest.setRequester(requester);
+        friendRequest.setReceiver(receiver);
+        friendRequest.setStatus(FriendRequestStatus.PENDING);
```
서비스 레이어 안에서가 아니라 FriendRequest 객체안에서 Update() 메소드를 통해 코드를 간략화가 가능해보입니다
EX -> friendRequest.update(requester,receiver);

anewsdev replied:
피드백 적용 완료 입니다.

seongjun1130 commented on src/main/java/com/sparta/newsfeedproject/domain/friend/entity/FriendRequest.java:
```diff
@@ -0,0 +1,31 @@
+    private Member receiver; // 친구 요청을 받은 유저
+
+    @Enumerated(EnumType.STRING)
+    private FriendRequestStatus status; //친구 요청의 상태 -> PENDING : 요청중
```
테이블 관계로 나뉘었을때 요청모음은 이미 테이블의 단위기에 status 가 필요한가 생각이 들지만 확장성 면에서는 나쁘지않을거 같습니다

anewsdev replied:
현재 프로젝트에서 확장 계획 없음에 따라 삭제 처리 하였습니다.

seongjun1130 commented on src/main/java/com/sparta/newsfeedproject/domain/friend/controller/FriendController.java:
```diff
@@ -0,0 +1,30 @@
+
+    //친구 요청 API
+    @PostMapping
+    public ResponseEntity<String> sendFriendRequest(@RequestBody FriendRequestDto requestDto) {
```
@LoginUser Member member 를 통해 로그인된 유저를 가져오고 신청하고자 하는 ID 는 @PathVariable 형태로 받아도 될거같아요

anewsdev replied:
피드백 적용 완료 입니다.

seongjun1130 reviewed (commented):
세부적인 예외만 좀 더 수정하시면 더 좋은코드가 될거같습니다! 조금 더 화이팅

upotato200 commented:
이 파트도 어려워보이네요 코드 잘 보았습니다, 어노테이션 관련해서 @getter, @setter 둘다 포함한게 아마 @data 로 기억해요 하나로 쓰셔도 좋을듯!

anewsdev committed:
fix : Add Friend Request
1.친구요청 예외처리
- 본인에게 친구요청
- 이미 친구인 경우 친구요청
- 이미 요청 상태인 경우 친구요청
2. service 간소화
3. status 삭제
4. @LoginUser 사용

anewsdev committed:
fix : update Friend
1. 불필요 변수 제거
2. 예외처리 메서드로 정리

seongjun1130 commented on src/main/java/com/sparta/newsfeedproject/domain/friend/service/FriendService.java:
```diff
@@ -26,33 +25,35 @@ public void sendFriendRequest(Long requesterId, Long receiverId) {
         Member receiver = memberRepository.findById(receiverId)
                 .orElseThrow(() -> new CustomException(ErrorCode.RECEIVER_NOT_FOUND));
```
명시적으로 누군가를 찾는지 예외를 나눠어주는것도 좋은방법같습니다.
코드재사용성을 생각하면 USER_NOT_FOUND 를 같이 공유해도 큰문제는없어보입니다.

anewsdev replied:
피드백 적용 완료입니다.

seongjun1130 commented on src/main/java/com/sparta/newsfeedproject/domain/friend/service/FriendService.java:
```diff
@@ -26,33 +25,35 @@ public void sendFriendRequest(Long requesterId, Long receiverId) {
+        friendRequestRepository.save(friendRequest);
+    }
+
+    private void validateFriendRequest(Long requesterId, Long receiverId, Member requester, Member receiver) {
```
ID 값들을 굳이 파라미터로 받지않아도 내부에서 Member requester, Member receiver 의 ID를 get으로 받아서 사용해도 가능하지않을까요??

anewsdev committed:
fix : update Friend
1. 예외코드 변경

anewsdev committed:
fix : update Friend
1. 오타 수정

anewsdev committed:
Merge branch 'member' into feat-2
